design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+        }
+      ]
+    }
+  }
+}
+```
+
+It lists the two assets (one file asset and one image asset) and, for each, it lists the publishing locations which include repository, key and publishing IAM role to assume.
+
+### Mutation
+
+The first stage in our pipeline is the mutation stage. This stage will include a single CodeBuild action that will simply execute:
+
+```console
+$ cdk deploy pipeline-*
+```
+
+This will update all the stacks with names that begin with "pipeline-". Namely, it includes the pipeline stack itself, and the auxiliary stacks that include the pipeline's regional replication buckets. Since we are using `cdk deploy` here, we can technically deploy any stack to any environment in this stage because `cdk deploy` can assume the deployment role in any of the environments which trust the deployment account.
+
+Notice that this stage happens **before** the publish stage. This is because the structure of the publish stage is dependent on which assets the app uses (we synthesize an action per asset for maximal parallelism), so we want to make sure the pipeline will be updated before publishing.
+
+> NOTE: if there is a constraint that does not allow the pipeline to be updated before publishing, it simply means that we have to publish all assets from a single action (`cdk-assets publish cdk.out`).
+
+In our example, there are 3 pipeline stacks:
+
+1. `pipeline-main`: contains the pipeline resource, the main artifacts bucket and other related resources
+2. `pipeline-us-east-1`: contains the pipeline replication bucket and key for us-east-1
+3. `pipeline-eu-west-2`: contains the pipeline replication bucket and key for eu-west-2
+
+Once we deploy these stacks, our pipeline will be ready to deploy the rest of our app.
+
+### Publishing
+
+The next stage in the process is the publishing stage.
+
+In our example, this stage will consist of two CodeBuild actions that will run the following commands concurrently. These command will package & upload the asset to all the environments. It will consult `assets.json` to determine the exact location into which to publish each asset and which cross-account role to assume.
+
+The first action will run this command:
+
+```shell
+cdk-assets publish cdk.out d31ca1aef8d1b68217852e7aea70b1e857d107b47637d5160f9f9a1b24882d2a
+```
+
+This will build the docker image from `cdk.out/my-image` using `CustomDockerFile` as a dockerfile. Then, it will assume the roles in the destinations specified in `assets.json` and push the image to the specified ECR locations.
+
+The second action will run this command:
+
+```shell
+cdk-assets publish cdk.out a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57
+```
+
+This will create a zip archive from the files under `asset.a0bae29e7b47044a66819606c65d26a92b1e844f4b3124a5539efc0167a09e57`, and then will assume the roles and upload the file to the destination S3 locations.
+
+### Deployment
```

Comment:
In your example, the service stacks have a dependency on the VPC stacks. How do we make sure the CI/CD deployment respects that dependency order?

Reply:
In the customer code this is implicit and this will be available to the app delivery APIs through `stack.dependencies`.